bigquery_etl/backfill/parse.py

```diff
@@ -0,0 +1,175 @@
+
+    @excluded_dates.validator
+    def validate_excluded_dates(self, attribute, value):
+        """Check that provided excluded dates are valid."""
+        if not all(map(lambda e: self.start_date < e < self.end_date, value)):
+            raise ValueError(f"Invalid excluded dates: {value}.")
+
+    @watchers.validator
+    def validate_watchers(self, attribute, value):
+        """Check that provided watchers are valid."""
+        if not value or not all(
+            map(lambda e: e and is_email_or_github_identity(e), value)
+        ):
+            raise ValueError(f"Invalid email or Github identity for watchers: {value}.")
+
+    @status.validator
+    def validate_status(self, attribute, value):
+        """Check that provided status is valid."""
+        if not hasattr(BackfillStatus, value.name):
+            raise ValueError(f"Invalid status: {value.name}.")
+
+    @staticmethod
+    def is_backfill_file(file_path: Path) -> bool:
+        """Check if the provided file is a backfill file."""
+        return os.path.basename(file_path) == BACKFILL_FILE
+
+    @classmethod
+    def entries_from_file(cls, file: Path) -> list:
+        """
+        Parse all backfill entries from the provided yaml file.
+
+        Return a list with all backfill entries.
+        """
+        if not cls.is_backfill_file(file):
+            raise ValueError(f"Invalid file: {file}.")
+
+        backfill_entries: List[Backfill] = []
+
+        with open(file, "r") as yaml_stream:
+            try:
+                backfills = yaml.load(yaml_stream, Loader=UniqueKeyLoader) or {}
+
+                for entry_date, entry in backfills.items():
+                    excluded_dates = []
+                    if "excluded_dates" in entry:
+                        excluded_dates = entry["excluded_dates"]
+
+                    backfill = cls(
+                        entry_date=entry_date,
+                        start_date=entry["start_date"],
+                        end_date=entry["end_date"],
+                        excluded_dates=excluded_dates,
+                        reason=entry["reason"],
+                        watchers=entry["watchers"],
+                        status=BackfillStatus[entry["status"].upper()],
+                    )
+
+                    backfill_entries.append(backfill)
+
+            except yaml.YAMLError as e:
+                raise e
+
+            return backfill_entries
+
+    def to_yaml(self) -> str:
+        """Create dictionary version of yaml for writing to file."""
+        yaml_dict = {
+            self.entry_date: {
+                "start_date": self.start_date,
+                "end_date": self.end_date,
+                "excluded_dates": sorted(self.excluded_dates),
+                "reason": self.reason,
+                "watchers": self.watchers,
+                "status": self.status.value,
+            }
+        }
+
+        if yaml_dict[self.entry_date]["excluded_dates"] == []:
```

Comment:
Is this actually necessary? backfill should ignore these dates if passed, but otherwise it should probably be able to handle if no values are provided?

Reply:
The purpose of this is so the "excluded_dates" field don't show up in the entry in the backfill.yaml file when it is empty.